README.rst
.. -*- coding: utf-8 -*-

dwdat2py
========

DEWESoft, a company making hardware for data acquisition, also provide
software applications (Windows) for working with data sets produced with
the hardware. Files are in their proprietary formats such as .d7d, .d7z
or .dxd. To work with the same data using a programming language like
Python it is necessary to first convert them to a portable format such
as text. To get around this extra step, DEWESoft provide a shared
library called DWDataReaderLib.

dwdat2py is a python package with wrappers around the functions provided
in DWDataReaderLib.

Installation
------------

Using pip
.........

::

   pip install dwdat2py

Installing from source
......................

Clone the repo with ``git clone https://github.com/tomnor/dwdat2py.git``, cd
into dwdat2py directory (the directory with the ``setup.py`` file) and say

::

   pip install .

Libraries are shipped with dwdat2py
...................................

By installation of dwdat2py, the Dewesoft libraries are included. Be
aware that the libraries are gratis but non-free binary blobs. The
libraries come with an EULA that permits use and redistribution in
third party projects as far as I understand it.

Libraries are installed in a sub-directory to dwdat2py called
``libs``.

Libraries are retrieved from
https://dewesoft.com/download/developer-downloads. Find a record
"Dewesoft Data Reader Library" and click the download button to get a
zip file (DWDataReader.zip) with the libs. Or skip that, since they
are included with dwdat2py.

Usage
-----

.. code:: python

   from dwdat2py import wrappers as dw

   dw.init()
   dw.open_data_file('dat1')
   # ... other functions to get at data in dat1 ...
   dw.close_data_file()
   dw.open_data_file('dat2')
   # ... other functions to get at data in dat2 ...
   dw.close_data_file() # work is done for now
   dw.de_init()
   # optionally init() again for a new session.


The calls ``init()`` and ``de_init()`` must be called first and last
respectively. Make sure ``open_data_file(...)`` is called successfully before
``close_data_file(...)``, else don't call close_data_file.

The highest level function for now in this package is this:

.. code:: python

    def channel_reduced(channel, reduction, encoding=None):
        """Return a flat list of data for channel reduced to reduction.

        Parameters
        ----------

        channel : int or str
            Either the channel index or the channel name.

        reduction : int
            One of the following
            time_stamp = 0
            ave = 1
            min = 2
            max = 3
            rms = 4

        encoding : str
            encoding to pass to `get_channel_list()`, which see.

        Wraps:
            Nothing explicit. This is a support function to simplify getting
            reduced data from a channel.

        """

And if you happen to know the index or name of the channels, this function
should suffice to get at the data, channel by channel. Else you would need to
call some helper functions first to prepare for this call.

Access to the wrappers module is also provided as a context manager:

.. code:: python

    @contextmanager
    def wrappersimport(fn, fsencoding=None):
        """Provide context access to the wrappers module.

        Return a handle to the wrappers module in a context manager and file
        `fn` (str) opened for operations (a .dxd file for example).
        Initialization and deinitialization is provided by this context
        manager, as well as opening and closing the file.

        The file information resulting from opening the file is available as
        a module level variable `fileinfo`, (`handle.fileinfo`).

        The function that `wrappers.open_data_file` wraps require bytes as
        file name. `fsencoding` is used in the call to
        `wrappers.open_data_file` but is hopefully not necessary to specify
        since os.fsencode() is used by default.

        Example usage:

        >>> import dwdat2py
        >>> with dwdat2py.wrappersimport(fn) as wi:
        ...    print(wi.fileinfo)
        ...    chlist = wi.get_channel_list(encoding='latin1')
        ...    for chinfo in chlist:
        ...        # print the average values from each channel (1)
        ...        print(wi.channel_reduced(chinfo.index, 1))
        ...    # get the "time stamps" (0)
        ...    time = wi.channel_reduced(chlist[0].index, 0)

        As with importing the wrappers module in the standard way, this will
        fail if the shared library is not found.

        """

Using libs not shipped with dwdat2py
....................................

The path to the libs can be specified by setting a variable
``dwdat2py.DEWELIBDIR`` to the directory containing the libs:

.. code:: python

    >>> import dwdat2py
    >>> dwdat2py.DEWELIBDIR = '~/latestdwlibs'

for example. This setting must be done before loading a datafile. The
structure of the directory should look like this:

.. code:: shell

    $ tree latestdwlibs
    latestdwlibs
    |-- DWDataReaderLib.dll
    |-- DWDataReaderLib.so
    |-- DWDataReaderLib64.dll
    `-- DWDataReaderLib64.so

Meaning, the shared libraries should sit directly inside the
``DEWELIBDIR`` directory. Only one of those files is really necessary
depending on the OS and platform.

Contribute
----------

Please report bugs and send suggestions or patches to the author. Or
make an issue or pull request on the repo home at `Github
<http://github.com/tomnor/dwdat2py>`_


CHANGES.rst
Changelog
=========

0.3.3 (2023-09-06)
------------------

- Libraries from Dewesoft updated. Dewesoft changelog say this::

    ## [v4_2_0_24] - 2023-01-13
    FIXED ISSUES:
      - Importing data files with unicode filenames does not work
      - DXD file through DWDataReader is not reading properly

    ## [v4_2_0_23] - 2022-06-09
    FEATURES, IMPROVEMENTS:
      - LongName for system monitor channels in DWDataReader library
      - Support async channel index in DWDataReader
      - Update python example in DWDataReder for block reading of IB buffers
      - Updating DWDatareader for faster reading of reduced data
    FIXED ISSUES:
      - Import of CNT-channels in analyze show wrong values
      - DER plugin doesn't import all samples

    ## [v4_2_0_22] - 2022-02-24
    FEATURES, IMPROVEMENTS:
    - Add helper functions for binary channels
    FIXED ISSUES:
    - Fix event time calculations

- Update of copyright years.
- Update readme on downloading the libraries.
- Update readme about using other libraries.


0.3.2 (2022-02-13)
------------------

- Libraries from Dewesoft updated. Dewesoft changelog say this::

    ## [v4_2_0_21] - 2021-11-29
    FIXED ISSUES:
    -Fix reading file store time in multifile
    -Support reading channels long name
    -DwExportHeader support for DXZ files
    -Fix opening DXZ files

    ## [v4_2_0_20] - 2021-08-02
    FEATURES, IMPROVEMENTS:
    - Improve Matlab example
    - Add mising functions to Linux example
    FIXED ISSUES:
    - Alarms and cursor events not working
    - Fix DWCloseDataFile function

- Update of copyright years.
- Prettify previous change note.


0.3.1 (2021-06-06)
------------------

- Libraries from Dewesoft updated. Dewesoft changelog say this::

    ## [v4_2_0_19] - 2021-06-04
    FIXED ISSUES:
    - DWGetScaledSamplesCount() is zero when using big data file with asynchronous channels


0.3.0 (2021-03-25)
------------------

- Libraries from Dewesoft included with dwdat2py installation.

- README.rst updated about the libs.

- Two new functions to get "scaled" samples, the raw full speed data
  samples, `get_scaled_samples_count()` and `get_scaled_samples()`.
  The latter respects `array_size` (array type of channels).

- A not documented function `get_storing_type()`
  (`DWGetStoringType()`) added with a module level dict `STORING_TYPE`
  to list the meaning of returned value.

- New functions `get_channel_factors`, `get_channel_props.`

- Array related functions `get_array_info_count`, `get_array_info_list`
  and `get_array_index_value` implemented.

- Added repo admin shell script utils/libadmin.sh to simplify commiting
  new version of the libraries. This script is now also responsible for
  the conversion of the doc file to text documentation.


0.2.0 (2020-06-21)
------------------

- A config file to specify path to the directory with library files can
  be used, (put in `~/.config` or `~`). `~` derived by
  `os.path.expanduser('~')`.

- Context manager provided for access to the wrappers module,
  `dwdat2py.wrappersimport()`, taking care of init, opening, closing
  data file and de_init.

- Written for Python 3.

- DWDataReaderHeader.py just copied as is to the repo. (instead of
  renaming to dwheader.py and tweaking it).

- Tests rewritten and more tests added.

- Change to Apache 2.0 license.

- DWDataReaderv.doc.txt is produced with

  ``$ antiword DWDataReaderv.doc > DWDataReaderv.doc.txt``

  when the doc file is updated.

0.1.0 (2017-01-01)
------------------

- Make public: Repo created


tests/test_init.py
"""
Test the __init__ module.
"""
import os
import sys
import unittest
import tempfile
import importlib

# Testing the local package
here = os.path.dirname(__file__)
packdir = os.path.abspath(os.path.join(here, os.pardir))
sys.path.insert(0, packdir)

import dwdat2py

DATAFILE1 = os.path.join(here, 'Example_Drive01.d7d')

if not os.path.exists(DATAFILE1):
    with gzip.open(DATAFILE1 + '.gz') as fi, open(DATAFILE1, 'wb') as fo:
        fo.write(fi.read())

# (time_stamp, ave, min, max, rms)
# keys are time stamp here
DATAFILE1RECORDS_CH0 = {
    0.0: (88.16272020339966, 86.67867183685303,
          89.28495049476624, 88.16669464111328),
    60.5: (82.5995683670044, 82.12530612945557,
           83.04696679115295, 82.59994506835938),
    95.5: (-0.08557390538044274, -0.09765923023223877,
           -0.06714072078466415, 0.0858997106552124)}


class TestWrappersImport(unittest.TestCase):

    def test_fileinfo(self):
        with dwdat2py.wrappersimport(DATAFILE1) as wi:
            self.assertEqual(wi.fileinfo.sample_rate, 100.0)
            # assertAlmostEqual is also available
            self.assertEqual(wi.fileinfo.start_store_time, 37903.8942918056)
            self.assertEqual(wi.fileinfo.duration, 95.8)

    def test_get_channel_list_count(self):

        with dwdat2py.wrappersimport(DATAFILE1) as wi:
            self.assertEqual(wi.get_channel_list_count(), 20)

    def test_get_channel_list(self):
        with dwdat2py.wrappersimport(DATAFILE1) as wi:
            channels = wi.get_channel_list()
            self.assertEqual(channels[0],
                             wi.Channel(index=0, name='GPSvel', unit='kph',
                                        description='v', color=16711680,
                                        array_size=1, data_type=2))
            self.assertEqual(channels[-1],
                             wi.Channel(index=27, name='CNT 0', unit='m',
                                        description='', color=65280,
                                        array_size=1, data_type=4))

    def test_channel_reduced_ave_by_index(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # wrappers.Channel(index=0, name=b'GPSvel', unit=b'kph',
        with dwdat2py.wrappersimport(DATAFILE1) as wi:
            averages = wi.channel_reduced(0, 1)
            # first and last time_stamp in DATAFILE1RECORDS_CH0 represent
            # the first and last record in channel 0
            self.assertEqual(DATAFILE1RECORDS_CH0[0.0][0], averages[0])
            self.assertEqual(DATAFILE1RECORDS_CH0[95.5][0], averages[-1])

    def test_channel_reduced_time_stamps_by_name(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # wrappers.Channel(index=0, name=b'GPSvel', unit=b'kph',
        with dwdat2py.wrappersimport(DATAFILE1) as wi:
            times = wi.channel_reduced('GPSvel', 0)
            self.assertEqual(min(times), 0)
            self.assertEqual(max(times), 95.5)
            for previous, current in zip(times, times[1:]):
                self.assertEqual(current - previous, 0.5)

    def test_channel_reduced_ave_by_name(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # wrappers.Channel(index=0, name=b'GPSvel', unit=b'kph',

        with dwdat2py.wrappersimport(DATAFILE1) as wi:
            averages = wi.channel_reduced('GPSvel', 1)
            # first and last time_stamp in DATAFILE1RECORDS_CH0 represent
            # the first and last record in channel 0
            self.assertEqual(DATAFILE1RECORDS_CH0[0.0][0], averages[0])
            self.assertEqual(DATAFILE1RECORDS_CH0[95.5][0], averages[-1])


if __name__ == '__main__':
    unittest.main()


tests/test_dwdat2py.py
#! /usr/bin/env python

"""Test the local dwdat2py package.
"""

import sys, os
import unittest
import gzip

# Testing the local package code but dependencies need to be available on the
# system.
here = os.path.dirname(__file__)
packdir = os.path.abspath(os.path.join(here, os.pardir))
sys.path.insert(0, packdir)

try:
    from dwdat2py import wrappers as dw
except EnvironmentError as e:
    print e.message
    print 'tests not possible without the lib, please see README.'
    sys.exit(1)

DATAFILE = os.path.join(here, 'Example_Drive01.d7d')

if not os.path.exists(DATAFILE):
    with gzip.open(DATAFILE + '.gz') as fi, open(DATAFILE, 'wb') as fo:
        fo.write(fi.read())


class AttrHolder:
    pass

# a global object to hold module attributes (file info, Channel list)
AH = AttrHolder()

class TestWrappers(unittest.TestCase):
    """Test all the functions in wrappers module."""

    # naming of tests is to ensure execution order
    def test01(self):
        """Test that init() return 0"""

        res = dw.init()
        self.assertEqual(res, 0)

    def test02(self):
        """Test open file and tuple (FileInfo) is returned."""
        AH.dw_fi = dw.open_data_file(DATAFILE)
        self.assertTrue(isinstance(AH.dw_fi, dw.FileInfo))

    def test03(self):
        """Attr's of FileInfo is familiar."""

        self.assertTrue(all([hasattr(AH.dw_fi, 'sample_rate'),
                             hasattr(AH.dw_fi, 'start_store_time'),
                             hasattr(AH.dw_fi, 'duration')]))

    def test04(self):
        """Check that version is higher than zero."""
        version = dw.get_version()
        self.assertGreater(version, 0)

    def test05(self):
        """Number of channels is 20."""
        numchannels = dw.get_channel_list_count()
        self.assertEqual(numchannels, 20)

    def test06(self):
        """Get a channel list and it's a list of Channel tuples"""

        AH.channels = dw.get_channel_list()
        self.assertTrue(all([isinstance(ct, dw.Channel) for ct in AH.channels]))

    def test07(self):
        """Attr's of a Channel in the channel list is familiar."""

        channel = AH.channels[0]
        truelist = [hasattr(channel, 'index'), hasattr(channel, 'name'),
                    hasattr(channel, 'unit'), hasattr(channel, 'description'),
                    hasattr(channel, 'color'), hasattr(channel, 'array_size'),
                    hasattr(channel, 'data_type')]

        self.assertTrue(all(truelist))

    def test08(self):
        """Reduced values count and sample time resolution are all the same."""

        countlist = [dw.get_reduced_values_count(ch.index)
                     for ch in AH.channels]

        self.assertTrue(all([cnt == (192, 0.5) for cnt in countlist]))

    def test09(self):
        """Each item type in the Channel tuples in the list is as expected."""

        truelist = []
        for channel in AH.channels:
            truelist += [type(channel.index) == int,
                         type(channel.name) in (str, unicode),
                         type(channel.unit) in (str, unicode),
                         type(channel.description) in (str, unicode),
                         type(channel.color) in (int, long),
                         type(channel.array_size) == int,
                         type(channel.data_type) == int]

        self.assertTrue(all(truelist))

    def test10(self):
        """get_reduced_values and channel_reduced results in the same data."""

        for channel in AH.channels:
            count, _ = dw.get_reduced_values_count(channel.index)
            records = dw.get_reduced_values(channel.index, 0, count)
            for reduction in range(5):
                self.assertEqual(dw.channel_reduced(channel.index, reduction),
                                     [record[reduction] for record in records])

    def test11(self):
        """The multiplexed argument to channel_reduced (int or str) works."""

        for channel in AH.channels:
            self.assertEqual(dw.channel_reduced(channel.index, 1),
                             dw.channel_reduced(channel.name, 1))


    def test20(self):
        """Test that close file return 0."""
        res = dw.close_data_file()
        self.assertEqual(res, 0)

    def shortDescription(self):
        return None



if __name__ == '__main__':
    unittest.main()


tests/test_wrappers.py
"""
Test the wrappers module.
"""
import sys
import os
import unittest
import gzip
from itertools import zip_longest
from collections import namedtuple

# Testing the local package code but dependencies need to be available on the
# system.
here = os.path.dirname(__file__)
packdir = os.path.abspath(os.path.join(here, os.pardir))
sys.path.insert(0, packdir)

try:
    from dwdat2py import wrappers
except EnvironmentError as e:
    print(e)
    print('tests not possible without the lib, please see README.')
    sys.exit(1)

DATAFILE1 = os.path.join(here, 'Example_Drive01.d7d')
DATAFILE2 = os.path.join(here, 'Test2.dxd')

if not os.path.exists(DATAFILE1):
    with gzip.open(DATAFILE1 + '.gz') as fi, open(DATAFILE1, 'wb') as fo:
        fo.write(fi.read())


# reduced channels
# (time_stamp, ave, min, max, rms)
# keys are time stamp here
DATAFILE1RECORDS_CH0 = {
    0.0: (88.16272020339966, 86.67867183685303,
          89.28495049476624, 88.16669464111328),
    60.5: (82.5995683670044, 82.12530612945557,
           83.04696679115295, 82.59994506835938),
    95.5: (-0.08557390538044274, -0.09765923023223877,
           -0.06714072078466415, 0.0858997106552124)}

DATAFILE2RECORDS_CH1 = {
    0.0: (0.0, 0.0, 0.0, 0.0),
    1.95: (0.0, 0.0, 0.0, 0.0),
    4.4: (0.0, 0.0, 0.0, 0.0)}

Counts = namedtuple('Counts', ('scaled', 'reduced'))

DATAFILE1COUNTS = {
    0: Counts(9580, 192),
    1: Counts(9580, 192),
    3: Counts(9557, 192),
    4: Counts(9557, 192),
    6: Counts(4791, 192),
    8: Counts(4791, 192),
    10: Counts(4791, 192),
    12: Counts(4791, 192),
    14: Counts(479, 192),
    16: Counts(4791, 192),
    17: Counts(4791, 192),
    18: Counts(4791, 192),
    19: Counts(4791, 192),
    21: Counts(4791, 192),
    22: Counts(191, 192),
    23: Counts(191, 192),
    24: Counts(1838, 192),
    25: Counts(1838, 192),
    26: Counts(191, 192),
    27: Counts(9580, 192),
}

DATAFILE1_TIMES_HEADS_TAILS = {
    # Selected channels ([:5], [-5:]) time_stamps of get_scaled_samples()
    0: ((0.0, 0.01, 0.02, 0.03, 0.04), (95.75, 95.76, 95.77, 95.78, 95.79)),
    3: ((0.0, 0.009999999776482582, 0.019999999552965164,
         0.029999999329447746, 0.03999999910593033),
        (95.75, 95.75999999046326, 95.77000001072884,
         95.7800000011921, 95.78999999165535)),
    6: ((0.0, 0.019999999552965164, 0.03999999910593033,
         0.05999999865889549, 0.07999999821186066),
        (95.70999999344349, 95.73000000417233, 95.75,
         95.77000001072884, 95.78999999165535)),
    14: ((0.18000000715255737, 0.3700000047683716, 0.5700000002980232,
          0.7700000107288361, 0.9699999988079071),
         (94.96000000834465, 95.15999999642372, 95.36000001430511,
          95.5599999986589, 95.75999999046326)),
    22: ((0.6400000005960464, 1.1400000005960464, 1.6400000005960464,
          2.1400000005960464, 2.6400000005960464),
         (93.64000000059605, 94.14000000059605, 94.64000000059605,
          95.14000000059605, 95.64000000059605)),
    24: ((0.6800000071525574, 0.7199999988079071, 0.7800000011920929,
          0.8199999928474426, 0.8799999952316284),
         (93.93999999761581, 93.97999998927116, 94.11999999731779,
          94.2800000011921, 94.31999999284744)),
}


class TestWrappersTestfile2(unittest.TestCase):

    def setUp(self):
        self.initresult = wrappers.init()
        if self.initresult != 0:
            self.stop()
        self.dwfileinfo = wrappers.open_data_file(DATAFILE2)

    def test_initresult(self):
        self.assertEqual(self.initresult, 0)

    def test_open_data_file_result(self):
        self.assertTrue(isinstance(self.dwfileinfo, wrappers.FileInfo))

    def test_get_storing_type(self):
        storeint = wrappers.get_storing_type()
        self.assertEqual(storeint, 0)
        self.assertEqual(wrappers.STORING_TYPE[storeint], 'ST_ALWAYS_FAST')

    def test_fileinfo_members(self):
        self.assertEqual(self.dwfileinfo.sample_rate, 20000.0)
        # assertAlmostEqual is also available
        self.assertEqual(self.dwfileinfo.start_store_time, 43998.6823785301)
        self.assertEqual(self.dwfileinfo.duration, 4.441)

    def test_get_version(self):
        version = wrappers.get_version()
        self.assertGreater(version, 0)
        self.assertIsInstance(version, int)

    def test_get_channel_list_count(self):
        self.assertEqual(wrappers.get_channel_list_count(), 2)

    def test_get_channel_list(self):
        channels = wrappers.get_channel_list()
        self.assertEqual(channels[0],
                         wrappers.Channel(index=0, name='Counting', unit='',
                                          description='Rainflow',
                                          color=16711680, array_size=20,
                                          data_type=5))
        self.assertEqual(channels[1],
                         wrappers.Channel(index=1, name='Peak/Valley',
                                          unit='', description='',
                                          color=16711680, array_size=1,
                                          data_type=5))

    def test_get_reduced_values_count(self):
        self.assertEqual(wrappers.get_reduced_values_count(0), (0, 0.05))
        self.assertEqual(wrappers.get_reduced_values_count(1), (89, 0.05))

    def test_get_reduced_values(self):
        # (time_stamp, ave, min, max, rms)
        records = wrappers.get_reduced_values(1, 0, 89)
        recdict = {record[0]: record[1:] for record in records}
        for timestamp, values in DATAFILE2RECORDS_CH1.items():
            for recvalue, should in zip_longest(recdict[timestamp], values):
                self.assertEqual(recvalue, should)

    def test_get_scaled_samples_count(self):
        self.assertEqual(wrappers.get_scaled_samples_count(1), 0)

    def test_get_scaled_samples(self):
        # DATAFILE2 has no normal scaled samples
        pass

    def test_channel_reduced_time_stamps_by_index(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # Channel(index=1, name=b'Peak/Valley', ...
        times = wrappers.channel_reduced(1, 0)
        self.assertEqual(min(times), 0)
        self.assertEqual(max(times), 4.4)
        for previous, current in zip(times, times[1:]):
            self.assertAlmostEqual(current - previous, 0.05)

    def test_channel_reduced_time_stamps_by_name(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # Channel(index=1, name=b'Peak/Valley', ...
        times = wrappers.channel_reduced('Peak/Valley', 0)
        self.assertEqual(min(times), 0)
        self.assertEqual(max(times), 4.4)
        for previous, current in zip(times, times[1:]):
            self.assertAlmostEqual(current - previous, 0.05)

    def test_channel_reduced_ave_by_name(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # Channel(index=1, name=b'Peak/Valley', ...
        averages = wrappers.channel_reduced('Peak/Valley', 1)
        # first and last time_stamp in DATAFILE2RECORDS_CH1 represent
        # the first and last record in channel 0
        self.assertEqual(DATAFILE2RECORDS_CH1[0.0][0], averages[0])
        self.assertEqual(DATAFILE2RECORDS_CH1[4.4][0], averages[-1])

    def test_channel_reduced_ave_by_name_ascii(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # Channel(index=1, name=b'Peak/Valley', ...
        averages = wrappers.channel_reduced('Peak/Valley', 1, encoding='ascii')
        # first and last time_stamp in DATAFILE2RECORDS_CH1 represent
        # the first and last record in channel 0
        self.assertEqual(DATAFILE2RECORDS_CH1[0.0][0], averages[0])
        self.assertEqual(DATAFILE2RECORDS_CH1[4.4][0], averages[-1])

    def tearDown(self):
        result = wrappers.close_data_file()
        if result:
            print('error: close_data_file() returned', result)
        result = wrappers.de_init()
        if result:
            print('error: de_init() returned', result)


class TestWrappersExampleFile01(unittest.TestCase):
    """Test all the functions in wrappers module."""

    def setUp(self):
        self.initresult = wrappers.init()
        if self.initresult != 0:
            self.stop()
        self.dwfileinfo = wrappers.open_data_file(DATAFILE1)

    def test_initresult(self):
        self.assertEqual(self.initresult, 0)

    def test_open_data_file_result(self):
        self.assertTrue(isinstance(self.dwfileinfo, wrappers.FileInfo))

    def test_get_storing_type(self):
        storeint = wrappers.get_storing_type()
        self.assertEqual(storeint, 0)
        self.assertEqual(wrappers.STORING_TYPE[storeint], 'ST_ALWAYS_FAST')

    def test_fileinfo_members(self):
        self.assertEqual(self.dwfileinfo.sample_rate, 100.0)
        # assertAlmostEqual is also available
        self.assertEqual(self.dwfileinfo.start_store_time, 37903.8942918056)
        self.assertEqual(self.dwfileinfo.duration, 95.8)

    def test_get_version(self):
        version = wrappers.get_version()
        self.assertGreater(version, 0)
        self.assertIsInstance(version, int)

    def test_get_channel_list_count(self):
        self.assertEqual(wrappers.get_channel_list_count(), 20)

    def test_get_channel_list(self):
        channels = wrappers.get_channel_list()
        self.assertEqual(channels[0],
                         wrappers.Channel(index=0, name='GPSvel', unit='kph',
                                          description='v', color=16711680,
                                          array_size=1, data_type=2))
        self.assertEqual(channels[-1],
                         wrappers.Channel(index=27, name='CNT 0', unit='m',
                                          description='', color=65280,
                                          array_size=1, data_type=4))

    def test_get_reduced_values_count(self):
        self.assertEqual(wrappers.get_reduced_values_count(0), (192, 0.5))
        self.assertEqual(wrappers.get_reduced_values_count(27), (192, 0.5))

    def test_get_reduced_values(self):
        # (time_stamp, ave, min, max, rms)
        records = wrappers.get_reduced_values(0, 0, 192)
        recdict = {record[0]: record[1:] for record in records}
        for timestamp, values in DATAFILE1RECORDS_CH0.items():
            for recvalue, should in zip_longest(recdict[timestamp], values):
                self.assertEqual(recvalue, should)

    def test_get_scaled_samples_count(self):
        # self.assertEqual(wrappers.get_scaled_samples_count(1), 0)
        chlist = wrappers.get_channel_list()
        for ch in chlist:
            count = wrappers.get_scaled_samples_count(ch.index)
            self.assertEqual(count, DATAFILE1COUNTS[ch.index].scaled)

    def test_get_scaled_samples(self):
        headtails = DATAFILE1_TIMES_HEADS_TAILS
        chlist = wrappers.get_channel_list()
        for ch in chlist:
            if ch.index in headtails:
                count = wrappers.get_scaled_samples_count(ch.index)
                time, data = wrappers.get_scaled_samples(ch.index, 0, count)
                self.assertEqual((time[:5], time[-5:]), headtails[ch.index])

    def test_channel_reduced_time_stamps_by_index(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # wrappers.Channel(index=0, name=b'GPSvel', unit=b'kph',
        times = wrappers.channel_reduced(0, 0)
        self.assertEqual(min(times), 0)
        self.assertEqual(max(times), 95.5)
        for previous, current in zip(times, times[1:]):
            self.assertEqual(current - previous, 0.5)

    def test_channel_reduced_ave_by_index(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # wrappers.Channel(index=0, name=b'GPSvel', unit=b'kph',
        averages = wrappers.channel_reduced(0, 1)
        # first and last time_stamp in DATAFILE1RECORDS_CH0 represent
        # the first and last record in channel 0
        self.assertEqual(DATAFILE1RECORDS_CH0[0.0][0], averages[0])
        self.assertEqual(DATAFILE1RECORDS_CH0[95.5][0], averages[-1])

    def test_channel_reduced_time_stamps_by_name(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # wrappers.Channel(index=0, name=b'GPSvel', unit=b'kph',
        times = wrappers.channel_reduced('GPSvel', 0)
        self.assertEqual(min(times), 0)
        self.assertEqual(max(times), 95.5)
        for previous, current in zip(times, times[1:]):
            self.assertEqual(current - previous, 0.5)

    def test_channel_reduced_ave_by_name(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # wrappers.Channel(index=0, name=b'GPSvel', unit=b'kph',
        averages = wrappers.channel_reduced('GPSvel', 1)
        # first and last time_stamp in DATAFILE1RECORDS_CH0 represent
        # the first and last record in channel 0
        self.assertEqual(DATAFILE1RECORDS_CH0[0.0][0], averages[0])
        self.assertEqual(DATAFILE1RECORDS_CH0[95.5][0], averages[-1])

    def test_channel_reduced_ave_by_name_ascii(self):
        #      0        1    2    3    4
        # (time_stamp, ave, min, max, rms)
        # wrappers.Channel(index=0, name=b'GPSvel', unit=b'kph',
        averages = wrappers.channel_reduced('GPSvel', 1, encoding='ascii')
        # first and last time_stamp in DATAFILE1RECORDS_CH0 represent
        # the first and last record in channel 0
        self.assertEqual(DATAFILE1RECORDS_CH0[0.0][0], averages[0])
        self.assertEqual(DATAFILE1RECORDS_CH0[95.5][0], averages[-1])

    def tearDown(self):
        result = wrappers.close_data_file()
        if result:
            print('error: close_data_file() returned', result)
        result = wrappers.de_init()
        if result:
            print('error: de_init() returned', result)


if __name__ == '__main__':
    unittest.main()


dwdat2py/DWDataReaderHeader.py

#----------------------------------------------------------------------------------------------------------------
# DWDataReader "header" for Python
#----------------------------------------------------------------------------------------------------------------
# Author: Dewesoft
# Notes:
#   - requires DWDataReaderLib.dll 4.0.0.0 or later
#   - tested with Python 3.4
#----------------------------------------------------------------------------------------------------------------

from ctypes import *
from enum import Enum
import sys

INT_SIZE = 4 # size of integer
DOUBLE_SIZE = 8 # size of double

class DWStatus(Enum):
    DWSTAT_OK = 0
    DWSTAT_ERROR = 1
    DWSTAT_ERROR_FILE_CANNOT_OPEN = 2
    DWSTAT_ERROR_FILE_ALREADY_IN_USE = 3
    DWSTAT_ERROR_FILE_CORRUPT = 4
    DWSTAT_ERROR_NO_MEMORY_ALLOC = 5
    DWSTAT_ERROR_CREATE_DEST_FILE = 6
    DWSTAT_ERROR_EXTRACTING_FILE = 7
    DWSTAT_ERROR_CANNOT_OPEN_EXTRACTED_FILE = 8
    DWSTAT_ERROR_INVALID_IB_LEVEL = 9

class DWChannelProps(Enum):
    DW_DATA_TYPE = 0
    DW_DATA_TYPE_LEN_BYTES = 1
    DW_CH_INDEX = 2
    DW_CH_INDEX_LEN = 3
    DW_CH_TYPE = 4
    DW_CH_SCALE = 5
    DW_CH_OFFSET = 6
    DW_CH_XML = 7
    DW_CH_XML_LEN = 8
    DW_CH_XMLPROPS = 9
    DW_CH_XMLPROPS_LEN = 10
    DW_CH_CUSTOMPROPS = 11
    DW_CH_CUSTOMPROPS_COUNT = 12
    DW_CH_LONGNAME = 13
    DW_CH_LONGNAME_LEN = 14

class DWChannelType(Enum):
    DW_CH_TYPE_SYNC = 0 # sync
    DW_CH_TYPE_ASYNC = 1 # async
    DW_CH_TYPE_SV = 2 # single value

class DWFileInfo(Structure):
    _pack_ = 1
    _fields_ =\
        [
            ("sample_rate", c_double),
            ("start_store_time", c_double),
            ("duration", c_double)
        ]

class DWChannel(Structure):
    _pack_ = 1
    _fields_ =\
        [
            ("index", c_int),
            ("name", c_char * 100),
            ("unit", c_char * 20),
            ("description", c_char * 200),
            ("color", c_uint),
            ("array_size", c_int),
            ("data_type", c_int)
        ]

class DWEvent(Structure):
    _pack_ = 1
    _fields_ =\
        [
            ("event_type", c_int),
            ("time_stamp", c_double),
            ("event_text", c_char * 200)
        ]

class DWReducedValue(Structure):
    _pack_ = 1
    _fields_ =\
        [
            ("time_stamp", c_double),
            ("ave", c_double),
            ("min", c_double),
            ("max", c_double),
            ("rms", c_double)
        ]

class DWArrayInfo(Structure):
    _pack_ = 1
    _fields_ =\
        [
            ("index", c_int),
            ("name", c_char * 100),
            ("unit", c_char * 20),
            ("size", c_int)
        ]

class DWCANPortData(Structure):
    _pack_ = 1
    _fields_ =\
        [
            ("arb_id", c_ulong),
            ("data", c_char * 8)
        ]

class DWComplex(Structure):
    _pack_ = 1
    _fields_ =\
        [
            ("re", c_double),
            ("im", c_double)
        ]

class DWEventType(Enum):
    etStart = 1
    etStop = 2
    etTrigger = 3
    etVStart = 11
    etVStop = 12
    etKeyboard = 20
    etNotice = 21
    etVoice = 22
    etModule = 24

class DWStoreType(Enum):
    ST_ALWAYS_FAST = 0
    ST_ALWAYS_SLOW = 1
    ST_FAST_ON_TRIGGER = 2
    ST_FAST_ON_TRIGGER_SLOW_OTH = 3

class DWDataType(Enum):
    dtByte = 0
    dtShortInt = 1
    dtSmallInt = 2
    dtWord = 3
    dtInteger = 4
    dtSingle = 5
    dtInt64 = 6
    dtDouble = 7
    dtLongword = 8
    dtComplexSingle = 9
    dtComplexDouble = 10
    dtText = 11
    dtBinary = 12
    dtCANPortData = 13
    dtCANFDPortData = 14
    dtBytes8 = 15
    dtBytes16 = 16
    dtBytes32 = 17
    dtBytes64 = 18

def DWRaiseError(err_str):
    print(err_str)
    sys.exit(-1)
